Global __class__ statement in class declaration leads to InvalidCastException

See the following code.  \_\_class\_\_ is  in class declaration which lead to an Unhandled Exception- InvalidCastException.


```
class A:
    global  __class__
    def a(self):
        super()
A()
```



Environmental: 

> IronPython 3.4.0a1 (3.4.0.0001)
> [.NETFramework,Version=v4.6 on Mono 6.12.0.122 (tarball Mon Feb 22 17:28:32 UTC 2021) (64-bit)] on linux
> 


```
(base) xxm@xxm-System-Product-Name:~$ ipy test.py

Unhandled Exception:
System.InvalidCastException: Specified cast is not valid.
at IronPython.Compiler.Ast.ClassDefinition.MakeClassBody () [0x00082] in <5cd2739cd27844619789a1aa1ffdb0b0>:0 
at IronPython.Compiler.Ast.ClassDefinition.GetLambda () [0x00008] in <5cd2739cd27844619789a1aa1ffdb0b0>:0 
at IronPython.Runtime.FunctionCode.get_Code () [0x00000] in <5cd2739cd27844619789a1aa1ffdb0b0>:0 
at IronPython.Runtime.FunctionCode.GetGeneratorOrNormalLambda () [0x0000b] in <5cd2739cd27844619789a1aa1ffdb0b0>:0 
at IronPython.Runtime.FunctionCode.UpdateDelegate (IronPython.Runtime.PythonContext context, System.Boolean forceCreation) [0x000d8] in <5cd2739cd27844619789a1aa1ffdb0b0>:0 
at IronPython.Compiler.Ast.ClassDefinition+<>c__DisplayClass64_0.<Reduce>b__0 (System.Object x) [0x00011] in <5cd2739cd27844619789a1aa1ffdb0b0>:0 
at System.Threading.QueueUserWorkItemCallback.WaitCallback_Context (System.Object state) [0x00007] in <533173d24dae460899d2b10975534bb0>:0 
at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00071] in <533173d24dae460899d2b10975534bb0>:0 
at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00000] in <533173d24dae460899d2b10975534bb0>:0 
at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem () [0x00021] in <533173d24dae460899d2b10975534bb0>:0 
at System.Threading.ThreadPoolWorkQueue.Dispatch () [0x00074] in <533173d24dae460899d2b10975534bb0>:0 
at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback () [0x00000] in <533173d24dae460899d2b10975534bb0>:0 
[ERROR] FATAL UNHANDLED EXCEPTION: System.InvalidCastException: Specified cast is not valid.
at IronPython.Compiler.Ast.ClassDefinition.MakeClassBody () [0x00082] in <5cd2739cd27844619789a1aa1ffdb0b0>:0 
at IronPython.Compiler.Ast.ClassDefinition.GetLambda () [0x00008] in <5cd2739cd27844619789a1aa1ffdb0b0>:0 
at IronPython.Runtime.FunctionCode.get_Code () [0x00000] in <5cd2739cd27844619789a1aa1ffdb0b0>:0 
at IronPython.Runtime.FunctionCode.GetGeneratorOrNormalLambda () [0x0000b] in <5cd2739cd27844619789a1aa1ffdb0b0>:0 
at IronPython.Runtime.FunctionCode.UpdateDelegate (IronPython.Runtime.PythonContext context, System.Boolean forceCreation) [0x000d8] in <5cd2739cd27844619789a1aa1ffdb0b0>:0 
at IronPython.Compiler.Ast.ClassDefinition+<>c__DisplayClass64_0.<Reduce>b__0 (System.Object x) [0x00011] in <5cd2739cd27844619789a1aa1ffdb0b0>:0 
at System.Threading.QueueUserWorkItemCallback.WaitCallback_Context (System.Object state) [0x00007] in <533173d24dae460899d2b10975534bb0>:0 
at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00071] in <533173d24dae460899d2b10975534bb0>:0 
at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00000] in <533173d24dae460899d2b10975534bb0>:0 
at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem () [0x00021] in <533173d24dae460899d2b10975534bb0>:0 
at System.Threading.ThreadPoolWorkQueue.Dispatch () [0x00074] in <533173d24dae460899d2b10975534bb0>:0 
at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback () [0x00000] in <533173d24dae460899d2b10975534bb0>:0 
```

--- a/Src/IronPython/Compiler/Ast/ClassDefinition.cs
+++ b/Src/IronPython/Compiler/Ast/ClassDefinition.cs
@@ -128,6 +128,7 @@ namespace IronPython.Compiler.Ast {
             if (TryGetVariable(reference.Name, out variable)) {
                 if (variable.Kind is VariableKind.Global) {
                     AddReferencedGlobal(reference.Name);
+                    return variable;
                 } else if (variable.Kind is VariableKind.Local) {
                     // TODO: This results in doing a dictionary lookup to get/set the local,
                     // when it should probably be an uninitialized check / global lookup for gets

--- a/Src/IronPython/Runtime/Operations/PythonOps.cs
+++ b/Src/IronPython/Runtime/Operations/PythonOps.cs
@@ -1498,7 +1498,7 @@ namespace IronPython.Runtime.Operations {
 
             // If __class__ is used, verify that it has been set
             if (vars._storage is RuntimeVariablesDictionaryStorage storage) {
-                int pos = Array.IndexOf(storage.Names, "__class__");
+                int pos = Array.IndexOf(storage.Names, "__class__", storage.NumFreeVars, storage.Names.Length - storage.NumFreeVars);
                 if (pos >= 0) {
                     ClosureCell classCell = storage.GetCell(pos);
                     if (!ReferenceEquals(classCell.Value, obj)) {

--- a/Src/IronPythonTest/Cases/CPythonCasesManifest.ini
+++ b/Src/IronPythonTest/Cases/CPythonCasesManifest.ini
@@ -887,7 +887,7 @@ Ignore=true
 Reason=ImportError: No module named _msi
 
 [CPython.test_super]
-Ignore=true
+Ignore=false # will have to be ignored again in 3.6
 
 [CPython.test_support]
 RunCondition=NOT $(IS_POSIX) # TODO: debug

